test/extended/openstack/topology.go

```diff
@@ -81,4 +90,234 @@ var _ = g.Describe("[sig-installer][Suite:openshift/openstack] The OpenShift clu
+			"enable_topology",
+		)
+
+		e2e.Logf("enable_topology: '%v'", enable_topology)
+
+		// Skip the test if enable_topology is not true
+		if enable_topology != "true" {
+			e2eskipper.Skipf("enable_topology needs to be true for this test to run")
+		}
+
+		// Set enable_topology value to false in the cloud provider config
+		g.By("Setting the enable_topology value to false")
+		err := setEnableTopologyKey(ctx, oc.AdminKubeClient(), false)
+		o.Expect(err).NotTo(o.HaveOccurred(), "Error setting the enable_topology param in the cloud-provider-config")
+
+		// Set the defer cleanup function for reverting the enable_topology change
+		g.DeferCleanup(func() {
+			ctx := context.Background()
+			enableTopologyCleanup(ctx, oc)
+		})
+
+		// Wait until enable_topology is changed in openshift-cluster-csi-drivers cloud-conf
+		g.By("Waiting until enable_topology is changed in the csi driver")
+		waitUntilEnableTopologyValue(ctx, oc, false)
+		e2e.Logf("enable_topology successfully changed in the csi driver")
+
+		// Wait until the openstack-cinder-csi-driver-node daemonset is redeployed
+		g.By("Waiting until the openstack-cinder-csi-driver-node daemonset is fully deployed and healthy")
+		waitUntilDaemonsetReady(ctx, oc, "openshift-cluster-csi-drivers", "openstack-cinder-csi-driver-node")
+		e2e.Logf("The openstack-cinder-csi-driver-node daemonset is fully deployed and healthy")
+
+		// Look for the default cinder-csi storage class
+		cinderSc := FindStorageClassByProvider(oc, "cinder.csi.openstack.org", true)
+		o.Expect(cinderSc).NotTo(o.BeNil(), "default cinder-csi storageClass not found.")
+
+		// Set the openstack volume client
+		g.By("Setting the openstack volume client")
+		volumeClient, err := client.GetServiceClient(ctx, openstack.NewBlockStorageV3)
+		o.Expect(err).NotTo(o.HaveOccurred(), "Failed to build the OpenStack client")
+
+		// Create a PVC
+		g.By("Creating a PVC")
+		ns := oc.Namespace()
+		pvc := CreatePVC(ctx, clientSet, "cinder-pvc", ns, cinderSc.Name, "1Gi")
+
+		// Create a deployment with the volume attached
+		g.By("Creating Openshift deployment with 1 replica and cinder volume attached")
+		labels := map[string]string{"app": "cinder-test-dep"}
+
+		testDeployment := createTestDeployment(deploymentOpts{
+			Name:     "cinder-test-dep",
+			Labels:   labels,
+			Replicas: 1,
+			Protocol: v1.ProtocolTCP,
+			Port:     8080,
+			Volumes: []volumeOption{{
+				Name:      "data-volume",
+				PvcName:   pvc.Name,
+				MountPath: "data",
+			}},
+		})
+
+		deployment, err := clientSet.AppsV1().Deployments(ns).Create(ctx,
+			testDeployment, metav1.CreateOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		err = e2edeployment.WaitForDeploymentComplete(clientSet, deployment)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// Get the PVCs from the test namespace
+		pvcs, err := GetPVCsFromNamespace(ctx, dc, ns)
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(len(pvcs)).To(o.Equal(1))
+
+		// Check the volume in openstack
```

Comment:
Should we have it in utils?

Reply:
Right, the functions related to volumes and PVS are defined in `volumes.go`. We can move them to `utils.go` but on the other hand we need to care about the size and structure of `utils.go` too.